src/github/mod.rs

```
@@ -235,6 +257,48 @@ impl DefaultGithubClient {
         // kick off the retry loop
         retry(Self::backoff_config(), operation).await
     }
+
+    async fn rate_limit_guard(&self) {
+        let mut state = self.rate_limit.lock().await;
+
+        // define a safety threshold
+        let threshold = 10;
+        if state.remaining <= threshold {
+            let now = Instant::now();
+            if now < state.reset_at {
+                let wait = state.reset_at - now;
+                tracing::info!(
+                    "Approaching rate limit ({} left). Sleeping {:?} until reset...",
+                    state.remaining,
+                    wait
+                );
+                tokio::time::sleep(wait).await;
```

When we have multiple tasks waiting and the rate limit resets, they may all send requests at once, potentially triggering secondary rate limits. We need to add jitter to `wait`

---

Hi @isSerge , these major flaws in my approach were fixed. Please do check my new approaches. Meanwhile I am writing more tests and fixing the hardcoded values and parsing error handling.